torch/csrc/autograd/engine.cpp

```diff
@@ -863,15 +863,33 @@ auto Engine::execute(const edge_list& roots,
       /* depth */ not_reentrant_backward_call ? 0 : total_depth + 1,
       /* cpu_ready_queue */ local_ready_queue);
 
+  // If we receive a single root, skip creating extra root node
+  bool skip_dummy_node = roots.size() == 1;
+  auto graph_root = skip_dummy_node ?
+    roots.at(0).function :
+    std::make_shared<GraphRoot>(roots, inputs);
+
   // Now compute the dependencies for all executable functions and queue the root
-  auto graph_root = std::make_shared<GraphRoot>(roots, inputs);
   compute_dependencies(graph_root.get(), *graph_task);
 
   if (!outputs.empty()) {
     graph_task->init_to_execute(*graph_root, outputs, accumulate_grad);
   }
 
-  execute_with_graph_task(graph_task, graph_root);
+  if (skip_dummy_node) {
+    auto inputs_ = std::vector<Variable>();
+    inputs_.reserve(graph_root->num_inputs());
+    for (size_t i = 0; i < graph_root->num_inputs(); ++i) {
+      if (i == roots.at(0).input_nr) {
+        inputs_.emplace_back(inputs.at(0));
```

Comment:
Shouldn't this be `at(i)`?

Reply:
Possible bug - when we have two inputs (e.g. for eig, fn->num_inputs() is 2 by definition), however, if we backprop from only the second input (input_nr = 1),  we have inputs.size() = 1. So, here doing at(i) would result in out of bounds access.
```
x = torch.tensor([[1., 0.], [0., 1.]], requires_grad=True)
y = torch.eig(x, eigenvectors=True)[1]

torch.autograd.grad(y, [x], torch.tensor([[1., 0.], [0., 1.]]))
```

Reply:
Ho right my bad.